File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-8.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(p.val>q.val){
            var temp = p;
            p = q;
            q = temp;
        }
        return findAncestor(root,p.val,q.val);
    }
    public TreeNode findAncestor(TreeNode node,int p, int q){
        if(node.val>=p && node.val<=q){
            return node;
        }
        if(q<node.val){
            return findAncestor(node.left,p,q);
        }
        return findAncestor(node.right,p,q);
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-8.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        self.res = -1001
        def dfs(cur):
            if not cur:
                return 0
            left = dfs(cur.left)
            right = dfs(cur.right)
            self.res = max(left+cur.val+right,cur.val,cur.val+right,cur.val+left,self.res)
            return max(left+cur.val,cur.val,cur.val+right)
        dfs(root)
        return self.res
